File: Go/ReverseVowels/reverseVowels.go
package main

import (
	"strings"
)

func swapChars(runes []rune, i, j int) []rune {
	tmp := runes[i]
	runes[i] = runes[j]
	runes[j] = tmp
	return runes
}

func reverseVowels(s string) string {
    vowelIndexes := []int{}
	for index, rune := range s {
		if strings.ContainsRune("aeiouAEIOU", rune) {
			vowelIndexes = append(vowelIndexes, index)
		}
	}
	runes := []rune(s)
	for i := 0; i < (len(vowelIndexes) / 2); i++ {
		runes = swapChars(runes, vowelIndexes[i], vowelIndexes[len(vowelIndexes) - 1 - i])
	}
	return string(runes)
}

File: TS/ReverseVowels/reverseVowels.ts
function reverseVowels(s: string): string {
  const vowelIndexes: number[] = [];
  const sArr: string[] = s.split("");
  Array.from(s).forEach((vowel, index) => {
    if ("aeiouAEIOU".includes(vowel)) {
      vowelIndexes.push(index);
    }
  });
  for (let tmp = "", i = 0; i < Math.floor(vowelIndexes.length / 2); i++) {
    // Swap vowels
    const leftVowelIndex = vowelIndexes[i];
    const rightVowelIndex = vowelIndexes[vowelIndexes.length - 1 - i];
    const leftVowel = sArr[leftVowelIndex];
    sArr[leftVowelIndex] = sArr[rightVowelIndex];
    sArr[rightVowelIndex] = leftVowel;
  }
  return sArr.join("");
}

export default reverseVowels;
